iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1110 @@
             }
-            finally
+            catch (MqttConnectingFailedException cfe)
             {
-                if (Logging.IsEnabled)
-                    Logging.Exit(this, cancellationToken, nameof(OpenAsync));
+                var connectCode = cfe.ResultCode;
+                switch (connectCode)
+                {
+                    case MqttClientConnectResultCode.BadUserNameOrPassword:
+                    case MqttClientConnectResultCode.NotAuthorized:
+                    case MqttClientConnectResultCode.ClientIdentifierNotValid:
+                        throw new IotHubClientException("Failed to open the MQTT connection due to incorrect or unauthorized credentials", false, IotHubStatusCode.Unauthorized, cfe);
+                    case MqttClientConnectResultCode.UnsupportedProtocolVersion:
+                        // Should never happen since the protocol version (3.1.1) is hardcoded
+                        throw new IotHubClientException("Failed to open the MQTT connection due to an unsupported MQTT version", cfe);
+                    case MqttClientConnectResultCode.ServerUnavailable:
+                        throw new IotHubClientException("MQTT connection rejected because the server was unavailable", true, IotHubStatusCode.ServerBusy, cfe);
+                    default:
+                        // MQTT 3.1.1 only supports the above connect return codes, so this default case
+                        // should never happen. For more details, see the MQTT 3.1.1 specification section "3.2.2.3 Connect Return code"
+                        // https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html
+                        // MQTT 5 supports a larger set of connect codes. See the MQTT 5.0 specification section "3.2.2.2 Connect Reason Code"
+                        // https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901074
+                        throw new IotHubClientException("Failed to open the MQTT connection", cfe);
+                }
             }
         }
 
         public override async Task SendEventAsync(Message message, CancellationToken cancellationToken)
         {
+            cancellationToken.ThrowIfCancellationRequested();
+
             try
             {
-                if (Logging.IsEnabled)
-                    Logging.Enter(this, cancellationToken, nameof(SendEventAsync));
+                // Topic depends on if the client is a module client or a device client
+                string baseTopicName = _moduleId == null ? _deviceToCloudMessagesTopic : _moduleToCloudMessagesTopic;
+                string TopicName = PopulateMessagePropertiesFromMessage(baseTopicName, message);
+
+                if (message.HasPayload && message.Payload.Length > MaxMessageSize)
+                {
+                    throw new InvalidOperationException($"Message size ({message.Payload.Length} bytes) is too big to process. Maximum allowed payload size is {MaxMessageSize}");
```

Comment:
The max message size here is what IoT hub defines. However, they also include other metadata in that calculation, which means there's a small chance we're allowing some messages that hub would eventually refuse. Should we remove any validation from our client and delegate all that to the service directly?

Reply:
I'd like to do that, sure.